lib/persistence.js

```diff
@@ -183,6 +184,55 @@ class Persistence {
     return { data: tdata, indexes: indexes }
   }
 
+  /**
+   * From a database's raw stream, return the corresponding
+   * machine understandable collection
+   */
+  treatRawStream (rawStream, cb) {
+    const dataById = {}
+    const tdata = []
+    const indexes = {}
+    let corruptItems = 0
+
+    const lineStream = byline(rawStream)
+    const that = this
+    let length = 0
+
+    lineStream.on('data', function(line) {
+
+      try {
+        const doc = model.deserialize(that.beforeDeserialization(line))
+        if (doc._id) {
+          if (doc.$$deleted === true) delete dataById[doc._id]
+          else dataById[doc._id] = doc
+        } else if (doc.$$indexCreated && doc.$$indexCreated.fieldName != null) indexes[doc.$$indexCreated.fieldName] = doc.$$indexCreated
+        else if (typeof doc.$$indexRemoved === 'string') delete indexes[doc.$$indexRemoved]
+      } catch (e) {
+        corruptItems += 1
+      }
+
+      length++
+    })
+
+    lineStream.on('end', function() {
+      // A bit lenient on corruption
+      let err
+      if (length > 0 && corruptItems / length > that.corruptAlertThreshold) {
+        err = new Error("More than " + Math.floor(100 * that.corruptAlertThreshold) + "% of the data file is corrupt, the wrong beforeDeserialization hook may be used. Cautiously refusing to start NeDB to prevent dataloss")
+      }
+
+      Object.keys(dataById).forEach(function (k) {
+        tdata.push(dataById[k])
+      })
+
+      cb(err, { data: tdata, indexes: indexes })
```

Comment:
I'm not entirely comfortable with this pattern of :
1. returning a callback with both an error and a result, this is something rather unconventional in node I think;
2. continuing to read the file even though the `corruptAlertThreshold` has been reached.

I'd rather call `cb(err, null)` when there is an error, and `cb(null, res)` when there is a result. An interesting item that could be returned here (but this would be another PR) could be the `corruptItems` count.

Reply:
I'm not certain about the logic behind some of these design decisions. As you know, I was not the original author of this pull request - I only ported the changes, improving things as I found it. I agree with both points:

- Returning both error and result is strange, normally one or the other is `null`
- It doesn't check the threshold while reading the stream, but instead it continues to read until the end

I made a change to address the first point. However, that causes one of the tests to fail, with `Cannot read property 'data' of null`. It's trying to access `result.data` when result is `null`. This is the test on which you commented below, which had a statement `console.log(err)`. I was going to change it to `assert.isNull(err)`, but then realized something about this test.

The test is called `Badly formatted lines have no impact on the treated data (with stream)`. There is the original non-stream version of the test above it. What these tests are doing, is to feed corrupt data and confirm that it's still possible to get the result ("treatedData").

What's interesting to note: the original non-stream test is passing, which means even with corrupt data, it's not throwing an error but returning a result. The stream version of the test is failing, and that console statement was logging: `Error: More than 10% of the data file is corrupt, the wrong beforeDeserialization hook may be used. Cautiously refusing to start NeDB to prevent dataloss`. Then the test continues to try accessing `result.data` to confirm that treated data still exists. When I changed the function to return `null` on error, it caused this test to fail.

OK, I made another commit `Return treated data even after corrupt threshold`. Now all tests are passing. I believe this follows the original intention for both the streaming and non-stream tests, that the treated data is returned even after corrupt threshold is reached.

Reply:
Hi @eliot-akira,
After digging a bit deeper, it appears there is a difference of behavior between `treatRawData` and `treatRawStream` in counting the `corruptItems`.

In the `treatRawData`'s implementation, the `corruptItems` counter starts at `-1` instead of `0` to ensure there are not false positive when files end with a blank line.

The test "Badly formatted lines have no impact on the treated data" tricks this by not putting a blank line at the end of `rawData`, that's why the `corruptAlertThreshold` is not reached when there is only 1 badly formatted line.

The test for `treatRawStream`'s implementation, however, can't use this trick because `byline` skips empty lines..

I think we should keep treatRawStream as close as possible to treatRawData as both will coexist.

To do so, I suggest we :
1. initialize `byline` with this option `const lineStream = byline(rawStream, { keepEmptyLines: true })` which will keep the last blank line,  like the other implementation;
2. initialize `corruptItems` with `-1` like the other implementation (could you add back original comment `// Last line of every data file is usually blank so not really corrupt` on both implementations, I cleaned it by mistake);
3. not return the results when an error happens in `treatRawStream` ;
4. transform the `assert.isNotNull(err)` into an `assert.isNull(err)`.

What do you think?

Reply:
Sure, that sounds like a good solution. I didn't notice the difference between the streaming and non-streaming functions, but it makes sense now why the streaming test is reaching corrupt data threshold, while the non-streaming test is not. OK, I will implement the changes as you suggested.